src/app.rs

```diff
@@ -199,18 +204,106 @@ impl App {
         Ok(app)
     }
 
-    async fn initialize_tree(
+    async fn restore_or_initialize_tree(
         database: &Database,
         tree_depth: usize,
         dense_prefix_depth: usize,
         gc_threshold: usize,
         initial_leaf_value: Hash,
+        mmap_file_path: String,
     ) -> AnyhowResult<TreeState> {
         let mut mined_items = database.get_commitments_by_status(Status::Mined).await?;
+        mined_items.sort_by_key(|item| item.leaf_index);
+
+        if let Some(tree_state) = Self::get_cached_tree_state(
+            database,
+            tree_depth,
+            dense_prefix_depth,
+            gc_threshold,
+            &initial_leaf_value,
+            &mined_items,
+            &mmap_file_path,
+        )
+        .await?
+        {
+            info!("tree restored from cache");
+            return Ok(tree_state);
+        }
+        let tree_state = Self::initialize_tree(
+            database,
+            tree_depth,
+            dense_prefix_depth,
+            gc_threshold,
+            initial_leaf_value,
+            mined_items,
+            mmap_file_path,
+        )
+        .await?;
+        info!("tree initialization successful");
+        Ok(tree_state)
+    }
+
+    async fn get_cached_tree_state(
+        database: &Database,
+        tree_depth: usize,
+        dense_prefix_depth: usize,
+        gc_threshold: usize,
+        initial_leaf_value: &Hash,
+        mined_items: &Vec<TreeUpdate>,
+        mmap_file_path: &str,
+    ) -> anyhow::Result<Option<TreeState>> {
+        let mut last_index_in_dense: usize = 0;
+        let leftover_items = if mined_items.is_empty() {
+            vec![]
+        } else {
+            let max_leaf = mined_items.last().map(|item| item.leaf_index).unwrap();
+            // if the last index is greater then dense_prefix_depth, 1 << dense_prefix_depth
+            // should be the last index in restored tree
+            last_index_in_dense = std::cmp::min(max_leaf, 1 << dense_prefix_depth);
```

Comment:
Shouldn't it be `(1 << dense_prefix_depth) - 1`?

Reply:
Oh ok, I see. The code is correct I think, but the name is confusing. Can you rename it to something like `first_index_outside_dense`?

Reply:
Do you think it's feasible to write a test case for this? I'm fairly confident that this is correct but I'd be more comfortable with a test

Reply:
that's not correct naming as well since its not always outside of dense tree. It could be first_free_index or something like that?